### PR TITLE
Add support for UNION [DISTINCT] sql

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/998

Earlier version https://github.com/apache/arrow-datafusion/pull/1029

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by xudong963** [comment]
Use the way mentioned by @Dandandan @alamb, `UNION` is easy to implement! Thanks again, PTAL @Dandandan @alamb @houqp

File: datafusion/src/sql/planner.rs [hunk 1]
```diff
@@ -170,8 +170,21 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+                    self.project(
+                        LogicalPlanBuilder::from(union_plan)
+                            .aggregate(projection_expr, vec![])?
+                            .build()?,
+                        vec![Expr::Wildcard],
+                    )
+                }
                 _ => Err(DataFusionError::NotImplemented(format!(
-                    "Only UNION ALL is supported, found {}",
+                    "Only UNION ALL and UNION are supported, found {}",
```
Comment (Dandandan) [inline datafusion/src/sql/planner.rs:187]:
```suggestion
                    "Only UNION ALL and UNION [DISTINCT] are supported, found {}",
```

File: datafusion/src/sql/planner.rs [hunk 2]
```diff
@@ -3440,7 +3453,7 @@ mod tests {
         let sql = "SELECT order_id from orders EXCEPT SELECT order_id FROM orders";
         let err = logical_plan(sql).expect_err("query should have failed");
         assert_eq!(
-            "NotImplemented(\"Only UNION ALL is supported, found EXCEPT\")",
+            "NotImplemented(\"Only UNION ALL and UNION are supported, found EXCEPT\")",
```
Comment (Dandandan) [inline datafusion/src/sql/planner.rs:3456]:
```suggestion
            "NotImplemented(\"Only UNION ALL and UNION [DISTINCT] are supported, found EXCEPT\")",
```

**Comment by Dandandan** [comment]
This looks good @xudong963 thank you!

One suggestion I have is to also test the `UNION DISTINCT` syntax, to make sure it's supported and has the same result as just `UNION`.

File: datafusion/src/execution/dataframe_impl.rs [hunk 3]
```diff
@@ -214,6 +215,20 @@ impl DataFrame for DataFrameImpl {
             .build()?;
         Ok(Arc::new(DataFrameImpl::new(self.ctx_state.clone(), &plan)))
     }
+
+    fn distinct(&self) -> Result<Arc<dyn DataFrame>> {
```
[the remark this answers is not in the document]
Comment (xudong963) [reply datafusion/src/execution/dataframe_impl.rs:219]: good idea!

**Comment by xudong963** [comment]
all comments have fixed @Dandandan @houqp

File: datafusion/src/logical_plan/builder.rs [hunk 4]
```diff
@@ -291,6 +291,15 @@ impl LogicalPlanBuilder {
         Ok(Self::from(union_with_alias(self.plan.clone(), plan, None)?))
     }
 
+    /// apply union distinct
+    pub fn union_distinct(&self) -> Result<Self> {
```
Comment (alamb) [inline datafusion/src/logical_plan/builder.rs:295]:
```suggestion
    /// Apply deduplication: Only distinct (different) values are returned)
    pub fn distinct(&self) -> Result<Self> {
```

I suggest renaming this to just `distinct` as there is only one input, so I find the use of the term `union` somewhat confusing

File: datafusion/src/sql/planner.rs [hunk 5]
```diff
@@ -170,8 +170,16 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
                     let right_plan = self.set_expr_to_plan(right.as_ref(), None, ctes)?;
                     union_with_alias(left_plan, right_plan, alias)
                 }
+                (SetOperator::Union, false) => {
+                    let left_plan = self.set_expr_to_plan(left.as_ref(), None, ctes)?;
+                    let right_plan = self.set_expr_to_plan(right.as_ref(), None, ctes)?;
+                    let union_plan = union_with_alias(left_plan, right_plan, alias)?;
+                    LogicalPlanBuilder::from(union_plan)
```
Comment (alamb) [inline datafusion/src/sql/planner.rs:177]: ❤️

File: datafusion/tests/user_defined_plan.rs [hunk 6]
```diff
@@ -128,6 +128,8 @@ const QUERY1: &str = "SELECT * FROM sales limit 3";
 const QUERY: &str =
     "SELECT customer_id, revenue FROM sales ORDER BY revenue DESC limit 3";
 
+const QUERY2: &str = "SELECT customer_id, revenue FROM sales  UNION SELECT customer_id, revenue FROM sales ORDER BY revenue DESC limit 3";
```
Comment (alamb) [inline datafusion/tests/user_defined_plan.rs:131]: Can you please move this test into `tests/sql.rs`? I don't think `UNION` has anything to do with user defined plans.

**Review by alamb — COMMENTED** [review]
This is a pretty awesome PR @xudong963  -- thank you! I really like how the support just sort of comes together from the existing parts and pieces

**Review by alamb — COMMENTED** [review]
I think renaming the function on LogicalPlanBuilder and moving the test is all I think is necessary prior to merging this PR.

**Comment by xudong963** [comment]
Thanks for your great comments which exactly improve my engineering ability! @alamb @houqp @Dandandan 

all comments have fixed @alamb

**Comment by xudong963** [comment]
> I really like how the support just sort of comes together from the existing parts and pieces

Yes, me too. The process helps me know well existing parts!

**Review by alamb — APPROVED** [review]
Great job @xudong963 !

**Comment by Dandandan** [comment]
Thanks @xudong963 ! 👍

**Pull request merged**